pkg/logs/pipeline/pipeline.go

```diff
@@ -20,39 +20,36 @@ import (
 
 // Pipeline processes and sends messages to the backend
 type Pipeline struct {
-	InputChan chan *message.Message
-	processor *processor.Processor
-	sender    *sender.Sender
+	InputChan    chan *message.Message
+	processor    *processor.Processor
+	mainSender   *sender.Sender
+	backupSender *sender.Sender
 }
 
 // NewPipeline returns a new Pipeline
 func NewPipeline(outputChan chan *message.Message, processingRules []*config.ProcessingRule, endpoints *config.Endpoints, destinationsContext *client.DestinationsContext, diagnosticMessageReceiver diagnostic.MessageReceiver, serverless bool, pipelineID int) *Pipeline {
-	var destinations *client.Destinations
-	if endpoints.UseHTTP {
-		main := http.NewDestination(endpoints.Main, http.JSONContentType, destinationsContext, endpoints.BatchMaxConcurrentSend)
-		additionals := []client.Destination{}
-		for _, endpoint := range endpoints.Additionals {
-			additionals = append(additionals, http.NewDestination(endpoint, http.JSONContentType, destinationsContext, endpoints.BatchMaxConcurrentSend))
-		}
-		destinations = client.NewDestinations(main, additionals)
-	} else {
-		main := tcp.NewDestination(endpoints.Main, endpoints.UseProto, destinationsContext)
-		additionals := []client.Destination{}
-		for _, endpoint := range endpoints.Additionals {
-			additionals = append(additionals, tcp.NewDestination(endpoint, endpoints.UseProto, destinationsContext))
-		}
-		destinations = client.NewDestinations(main, additionals)
+	mainDestinations := getMainDestinations(endpoints, destinationsContext)
+	var backupDestinations *client.Destinations
+	if endpoints.Backup != nil {
+		backupDestinations = getBackupDestinations(endpoints, destinationsContext)
 	}
 
 	senderChan := make(chan *message.Message, config.ChanSize)
 
-	var strategy sender.Strategy
-	if endpoints.UseHTTP || serverless {
-		strategy = sender.NewBatchStrategy(sender.ArraySerializer, endpoints.BatchWait, endpoints.BatchMaxConcurrentSend, endpoints.BatchMaxSize, endpoints.BatchMaxContentSize, "logs", pipelineID)
+	var mainSender *sender.Sender
+	var backupSender *sender.Sender
+
+	// If there is a backup endpoint - we are dual-shipping so we need to spawn an additional sender.
+	if backupDestinations != nil {
+		mainSenderChannel := make(chan *message.Message, config.ChanSize)
+		backupSenderChannel := make(chan *message.Message, config.ChanSize)
+
+		mainSender = sender.NewSender(mainSenderChannel, outputChan, mainDestinations, getStrategy(endpoints, serverless, pipelineID))
+		backupSender = sender.NewSender(backupSenderChannel, outputChan, backupDestinations, getStrategy(endpoints, serverless, pipelineID))
+		sender.SplitSenders(senderChan, mainSender, backupSender)
 	} else {
-		strategy = sender.StreamStrategy
+		mainSender = sender.NewSender(senderChan, outputChan, mainDestinations, getStrategy(endpoints, serverless, pipelineID))
```

Comment:
nit: This is more of a design aspect thing but the if/else seems like it's selecting distinct and unique logic path based on if the backup is defined. While it's functionally fine, having a single flow through this block may catch errors early for less used options.

Effectively, I wonder if having a variadic `sender.SplitSenders(senderChan chan *message.Message, sourceChans ...*Sender)` and then deciding in that function if you want to have the extra buffering channel or not, would cause the whole flow to go through a single flow of logic and be more resilient to future regressions.

Either way, functionally there's nothing wrong with your implementation so feel free to ignore this.